File: articles/natural-selection-adaptation-fitness/r/quantitative_trait_selection.R
# Quantitative trait selection and breeder's equation response in R.

library(dplyr)

set.seed(42)

n <- 2000
trait <- rnorm(n, mean = 0, sd = 1)

fitness <- exp(0.6 * trait)
fitness <- fitness / mean(fitness)

population <- tibble(
  trait = trait,
  relative_fitness = fitness
)

mean_trait_before <- mean(population$trait)
selected_mean <- weighted.mean(population$trait, w = population$relative_fitness)
selection_differential <- selected_mean - mean_trait_before

heritability <- 0.45
response_to_selection <- heritability * selection_differential
predicted_next_mean <- mean_trait_before + response_to_selection

summary_tbl <- tibble(
  mean_before_selection = mean_trait_before,
  selected_weighted_mean = selected_mean,
  selection_differential_S = selection_differential,
  heritability_h2 = heritability,
  response_R = response_to_selection,
  predicted_next_generation_mean = predicted_next_mean
)

print(summary_tbl)

File: articles/natural-selection-adaptation-fitness/python/quantitative_trait_selection.py
"""
Quantitative trait selection and breeder's equation response.

Run:
    python python/quantitative_trait_selection.py
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def main() -> None:
    """Estimate selection differential and predicted response."""

    rng = np.random.default_rng(7)

    n = 5000
    trait = rng.normal(loc=0.0, scale=1.0, size=n)

    fitness = np.exp(0.5 * trait - 0.15 * trait**2)
    fitness = fitness / np.mean(fitness)

    mean_before = trait.mean()
    selected_mean = np.average(trait, weights=fitness)
    selection_differential = selected_mean - mean_before

    heritability = 0.40
    response = heritability * selection_differential
    predicted_next_mean = mean_before + response

    summary = pd.DataFrame(
        {
            "mean_before": [mean_before],
            "selected_mean": [selected_mean],
            "selection_differential_S": [selection_differential],
            "heritability_h2": [heritability],
            "response_R": [response],
            "predicted_next_mean": [predicted_next_mean],
        }
    )

    print(summary.round(4).to_string(index=False))


if __name__ == "__main__":
    main()
